units/distro_bleed_terminations.go

```diff
@@ -0,0 +1,181 @@
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    hostBleedJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	j.SetPriority(2)
+
+	return j
+}
+
+func NewHostBleedJob(env evergreen.Environment, bleedInfo BleedInfo, id string) amboy.Job {
+	j := makeHostBleedJob()
+	j.BleedInfo = bleedInfo
+	j.env = env
+	j.SetID(fmt.Sprintf("%s.%s", hostBleedJobName, id))
+	return j
+}
+
+func (j *hostBleedJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	if j.settings == nil {
+		j.settings = j.env.Settings()
+	}
+
+	if j.HasErrors() {
+		return
+	}
+
+	// get currently existing hosts, in case some hosts have already been terminated elsewhere
+	existingHostCount, err := host.CountRunningHosts(j.BleedInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting existing hosts by Distro.Id"))
+		return
+	}
+	idleHosts, err := host.IdleHostsWithDistroID(j.BleedInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting idle hosts by Distro.Id"))
+		return
+	}
+	bleedTarget := existingHostCount - j.BleedInfo.NewCapTarget
+
+	for i := 0; i < len(idleHosts); i++ {
+		j.AddError(j.checkAndTerminateHost(ctx, &idleHosts[i], &bleedTarget))
+		if bleedTarget == 0 {
+			break
+		}
+	}
+
+	grip.Info(message.Fields{
+		"id":                   j.ID(),
+		"job_type":             hostBleedJobName,
+		"distro_id":            j.BleedInfo.DistroID,
+		"num_idle_hosts":       len(idleHosts),
+		"num_terminated_hosts": j.Terminated,
+		"terminated_hosts":     j.TerminatedHosts,
+	})
+}
+
+func (j *hostBleedJob) checkAndTerminateHost(ctx context.Context, h *host.Host, bleedTarget *int) error {
+	if !h.IsEphemeral() {
+		grip.Notice(message.Fields{
+			"job":      j.ID(),
+			"host_id":  h.Id,
+			"job_type": j.Type().Name,
+			"status":   h.Status,
+			"provider": h.Distro.Provider,
+			"message":  "host termination for a non-ephemeral distro",
+			"cause":    "programmer error",
+		})
+		return errors.New("attempted to terminate non-ephemeral host")
+	}
+
+	// ask the host how long it has been idle
+	idleTime := h.IdleTime()
+
+	// if the communication time is > 10 mins then there may not be an agent on the host.
+	communicationTime := h.GetElapsedCommunicationTime()
+
+	// get a cloud manager for the host
+	mgrOpts, err := cloud.GetManagerOptions(h.Distro)
+	if err != nil {
+		return errors.Wrapf(err, "can't get ManagerOpts for host '%s'", h.Id)
+	}
+	manager, err := cloud.GetManager(ctx, j.env, mgrOpts)
+	if err != nil {
+		return errors.Wrapf(err, "error getting cloud manager for host %v", h.Id)
+	}
+
+	// ask how long until the next payment for the host
+	tilNextPayment := manager.TimeTilNextPayment(h)
+
+	if tilNextPayment > maxTimeTilNextPayment {
+		return nil
+	}
+
+	if h.IsWaitingForAgent() && (communicationTime < idleWaitingForAgentCutoff || idleTime < idleWaitingForAgentCutoff) {
```

Comment:
I would argue to do this check before 139 - 153; you don't need to get the manager if you're going to exit early anyway (or else calculate idleTime and communicationTime after checking payment, to make it clearer)